specs/data_structures.md

```diff
@@ -320,31 +286,178 @@ A proof into an SMT is structured as:
+- `SHARE_SIZE` bytes per piece
+
+[Leopard-RS](https://github.com/catid/leopard) is a C library that implements the above scheme with quasilinear runtime.
+
+## 2D Reed-Solomon Encoding Scheme
+
+The 2-dimensional data layout is described in this section. Data is arranged in `k * k` matrix. The roots of [NMTs](#namespace-merkle-tree) for each row and column across four quadrants of data, `Q0` to `Q3` must be computed. In other words, `2k` row roots and `2k` column roots must be computed. The row and column roots are stored in the `availableDataCommitments` of the [AvailableDataHeader](#availabledataheader).
+
+![fig: RS2D encoding: data quadrants.](figures/rs2d_quadrants.svg)
+
+The data of `Q0` is the original data, and the remaining quadrants are parity data. Where `A -> B` indicates that `B` is computed using [erasure coding](#reed-solomon-erasure-coding) from `A`:
+- `Q0 -> Q1` for each row in `Q0` and `Q1`
+- `Q0 -> Q2` for each column in `Q0` and `Q2`
+- `Q2 -> Q3` for each row in `Q2` and `Q3`
+
+![fig: RS2D encoding: extending data.](figures/rs2d_extending.svg)
+
+As an example, the parity data in the second column of `Q2` (in striped purple) is computed by [extending](#reed-solomon-erasure-coding) the original data in the second column of `Q0` (in solid blue).
+
+![fig: RS2D encoding: extending a column.](figures/rs2d_extend.svg)
+
+Now that all four quadrants are filled, the row and column roots can be computed. To do so, each row/column is used as the leaves of a [NMT](#namespace-merkle-tree), for which the compact root is computed (i.e. an extra hash operation is used to produce a single [HashDigest](#hashdigest)). In this example, the fourth row root value is computed as the NMT root of the fourth row of `Q0` and the fourth row of `Q1` as leaves.
+
+![fig: RS2D encoding: a row root.](figures/rs2d_row.svg)
+
+Finally, the `availableDataRoot` of the block [Header](#header) is computed as the Merkle root of the [binary Merkle tree](#binary-merkle-tree) with the row and column roots as leaves.
+
+![fig: Available data root.](figures/data_root.svg)
+
+## Share
+
+A share is a fixed-size data chunk that will be erasure-coded and committed to in [Namespace Merkle trees](#namespace-merkle-tree).
+
+| name      | type               | description     |
+| --------- | ------------------ | --------------- |
+| `rawData` | `byte[SHARE_SIZE]` | Raw share data. |
+
+An example layout of the share's internal bytes is shown below. For non-parity shares _with a reserved namespace_, the first `SHARE_RESERVED_BYTES` bytes (`*` in the figure) is the starting byte of the first request in the share as an unsigned integer, or `0` if there is none. In this example, the first byte would be `80` (or `0x50` in hex). For shares _with a non-reserved namespace_ (and parity shares), the first `SHARE_RESERVED_BYTES` bytes have no special meaning and are simply used to store data like all the other bytes in the share.
```

Comment:
> is the starting byte of the first request in the share as an unsigned integer,

Now I see what you meant with starting byte. I find the word request in this context confusing. Wouldn't it be simpler if all `tx` wher length prefixed? Then you'd know when tx2 will end and tx3 will start.

Reply:
I used "request" here in the context of the [Tendermint paper](https://arxiv.org/abs/1807.04938), which is equivalent to "transactions" for a distributed ledger. Probably not the best word. Any ideas for something that can describe "any entry in the ledger"?

> Wouldn't it be simpler if all tx were length prefixed? Then you'd know when tx2 will end and tx3 will start.
You mean just if every request in a share had a length prefix? I don't know if that would be necessarily be simpler, since the number of "length bytes" would be variable instead of fixed. With this scheme you can just take 255-byte slices of the serialized requests.

Reply:
> You mean just if every request in a share had a length prefix?

Yes (basically len(entry )|| entry, for each entry.

>  I don't know if that would be necessarily be simpler, since the number of "length bytes" would be variable instead of fixed. With this scheme you can just take 255-byte slices of the serialized requests.

Hmm, good point. Let's keep it as is. I'll implement this and if there is any insights from that we can still iterate in a separate PR.